Question: Doing a maven build for the "impl" package yields two failures (independent of the changes, they were there before)
```
[ERROR] Redundant nullcheck of connection, which is known to be non-null in org.tinylog.writers.JdbcWriterTest$AbstractTest.executeSql(String) [org.tinylog.writers.JdbcWriterTest$AbstractTest, org.tinylog.writers.JdbcWriterTest$AbstractTest] Redundant null check at JdbcWriterTest.java:[line 870]Another occurrence at JdbcWriterTest.java:[line 874] RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE
[ERROR] Nullcheck of statement at line 872 of value previously dereferenced in org.tinylog.writers.JdbcWriterTest$AbstractTest.executeSql(String) [org.tinylog.writers.JdbcWriterTest$AbstractTest, org.tinylog.writers.JdbcWriterTest$AbstractTest] At JdbcWriterTest.java:[line 872]Redundant null check at JdbcWriterTest.java:[line 873] RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE
```

Is this expected or do I build it wrong. This happends at the end and the packages are build correctly anyway so it does not matter too much.

--- a/tinylog-impl/src/main/java/org/tinylog/pattern/PackageNameToken.java
+++ b/tinylog-impl/src/main/java/org/tinylog/pattern/PackageNameToken.java
@@ -56,6 +56,9 @@ final class PackageNameToken implements Token {
 	 * @return Package name or {@code null} for default package
 	 */
 	private static String getPackage(final String fullyQualifiedClassName) {
+		if (fullyQualifiedClassName == null) {
+			return null;
+		}
 		int dotIndex = fullyQualifiedClassName.lastIndexOf('.');
 		if (dotIndex == -1) {
 			return null;

--- a/tinylog-impl/src/main/java/org/tinylog/pattern/SimpleClassNameToken.java
+++ b/tinylog-impl/src/main/java/org/tinylog/pattern/SimpleClassNameToken.java
@@ -57,6 +57,9 @@ final class SimpleClassNameToken implements Token {
 	 * @return Class name without package prefix
 	 */
 	private static String getSimpleClassName(final String fullyQualifiedClassName) {
+		if (fullyQualifiedClassName == null) {
+			return null;
+		}		
 		int dotIndex = fullyQualifiedClassName.lastIndexOf('.');
 		if (dotIndex < 0) {
 			return fullyQualifiedClassName;

--- a/tinylog-impl/src/main/java/org/tinylog/pattern/ThreadIdToken.java
+++ b/tinylog-impl/src/main/java/org/tinylog/pattern/ThreadIdToken.java
@@ -37,12 +37,14 @@ final class ThreadIdToken implements Token {
 
 	@Override
 	public void render(final LogEntry logEntry, final StringBuilder builder) {
-		builder.append(logEntry.getThread().getId());
+		Thread thread = logEntry.getThread();
+		builder.append(thread == null ? "<ID of thread is not set>" : thread.getId());
 	}
 	
 	@Override
 	public void apply(final LogEntry logEntry, final PreparedStatement statement, final int index) throws SQLException {
-		statement.setLong(index, logEntry.getThread().getId());
+		Thread thread = logEntry.getThread();
+		statement.setLong(index, thread == null ? -1 : thread.getId());
 	}
 
 }

--- a/tinylog-impl/src/main/java/org/tinylog/pattern/ThreadNameToken.java
+++ b/tinylog-impl/src/main/java/org/tinylog/pattern/ThreadNameToken.java
@@ -37,12 +37,14 @@ final class ThreadNameToken implements Token {
 
 	@Override
 	public void render(final LogEntry logEntry, final StringBuilder builder) {
-		builder.append(logEntry.getThread().getName());
+		Thread thread = logEntry.getThread();
+		builder.append(thread == null ? "<Thread is not set>" : thread.getName());
 	}
 	
 	@Override
 	public void apply(final LogEntry logEntry, final PreparedStatement statement, final int index) throws SQLException {
-		statement.setString(index, logEntry.getThread().getName());
+		Thread thread = logEntry.getThread();
+		statement.setString(index, thread == null ? "<Thread is not set>" : thread.getName());
 	}
 
 }

--- a/tinylog-impl/src/test/java/org/tinylog/pattern/PackageNameTokenTest.java
+++ b/tinylog-impl/src/test/java/org/tinylog/pattern/PackageNameTokenTest.java
@@ -73,6 +73,15 @@ public final class PackageNameTokenTest {
 	}
 
 	/**
+	 * Verifies that nothing will be appended to a {@link StringBuilder} for a null class name.
+	 */
+	@Test
+	public void renderClassWithoutPackageIfClassnameIsNull() {
+		PackageNameToken token = new PackageNameToken();
+		assertThat(render(token, null)).isEmpty();
+	}
+	
+	/**
 	 * Verifies that {@code null} will be added to a {@link PreparedStatement} for classes in the default package.
 	 *
 	 * @throws SQLException

--- a/tinylog-impl/src/test/java/org/tinylog/pattern/SimpleClassNameTokenTest.java
+++ b/tinylog-impl/src/test/java/org/tinylog/pattern/SimpleClassNameTokenTest.java
@@ -74,6 +74,15 @@ public final class SimpleClassNameTokenTest {
 	}
 
 	/**
+	 * Verifies that nothing will be appended to a {@link StringBuilder} for null class names.
+	 */
+	@Test
+	public void renderClassWithoutPackageIfClassnameIsNull() {
+		SimpleClassNameToken token = new SimpleClassNameToken();
+		assertThat(render(token, null)).isEqualTo("null");
+	}
+	
+	/**
 	 * Verifies that the simple name of a class will be added to a {@link PreparedStatement} for classes in the default
 	 * package.
 	 *

--- a/tinylog-impl/src/test/java/org/tinylog/pattern/ThreadIdTokenTest.java
+++ b/tinylog-impl/src/test/java/org/tinylog/pattern/ThreadIdTokenTest.java
@@ -48,6 +48,15 @@ public final class ThreadIdTokenTest {
 		Thread thread = new Thread();
 		assertThat(render(token, thread)).isEqualTo(String.valueOf(thread.getId()));
 	}
+	
+	/**
+	 * Verifies that the ID of the thread will be rendered correctly even of the thread is null.
+	 */
+	@Test
+	public void renderThreadIdIfThreadIsNull() {
+		ThreadIdToken token = new ThreadIdToken();
+		assertThat(render(token, null)).isEqualTo("<ID of thread is not set>");
+	}
 
 	/**
 	 * Verifies that the ID of the thread will be added to a {@link PreparedStatement}.
@@ -65,6 +74,21 @@ public final class ThreadIdTokenTest {
 		verify(statement).setLong(1, thread.getId());
 	}
 
+	/**
+	 * Verifies that the ID of the thread will be added to a {@link PreparedStatement} if thread is null.
+	 *
+	 * @throws SQLException
+	 *             Failed to add value to prepared SQL statement
+	 */
+	@Test
+	public void applyThreadIdIfThreadIsNull() throws SQLException {
+		ThreadIdToken token = new ThreadIdToken();
+
+		PreparedStatement statement = mock(PreparedStatement.class);
+		token.apply(createLogEntry(null), statement, 1);
+		verify(statement).setLong(1, -1);
+	}
+	
 	/**
 	 * Renders a token.
 	 *

--- a/tinylog-impl/src/test/java/org/tinylog/pattern/ThreadNameTokenTest.java
+++ b/tinylog-impl/src/test/java/org/tinylog/pattern/ThreadNameTokenTest.java
@@ -48,6 +48,15 @@ public final class ThreadNameTokenTest {
 		Thread thread = new Thread("MyThread");
 		assertThat(render(token, thread)).isEqualTo("MyThread");
 	}
+	
+	/**
+	 * Verifies that the name of the thread will be rendered correctly even if the thread is null.
+	 */
+	@Test
+	public void renderThreadNameIfThreadIsNull() {
+		ThreadNameToken token = new ThreadNameToken();
+		assertThat(render(token, null)).isEqualTo("<Thread is not set>");
+	}
 
 	/**
 	 * Verifies that the name of the thread will be added to a {@link PreparedStatement}.
@@ -65,6 +74,21 @@ public final class ThreadNameTokenTest {
 		verify(statement).setString(1, "MyThread");
 	}
 
+	/**
+	 * Verifies that the name of the thread will be added to a {@link PreparedStatement} if the thread is null.
+	 *
+	 * @throws SQLException
+	 *             Failed to add value to prepared SQL statement
+	 */
+	@Test
+	public void applyThreadNameIfThreadIsNull() throws SQLException {
+		ThreadNameToken token = new ThreadNameToken();
+
+		PreparedStatement statement = mock(PreparedStatement.class);
+		token.apply(createLogEntry(null), statement, 1);
+		verify(statement).setString(1, "<Thread is not set>");
+	}
+	
 	/**
 	 * Renders a token.
 	 *

